### PR TITLE
[coq] pick flags from profile

### LINKED ISSUE #3486: [wish] setting coq.theory.flags from a profile
It seems flags to `coqc` can only be set from a `coq.theory` stanza, while it would be nice to set them from profiles as well (as one can do for the options for `ocamlc`).

Cc @ejgallego

### PR DESCRIPTION
**Pull request by gares**
*disclaimer* I don't know the code of dune at all, this is symbol pushing

This PR adds `(coq (flags <osl>))` to profile

OUTDATED:
~This PR makes it possible to use different Coq compilation backends via profiles. Today Coq can produce object files in 3 ways:
- complete check (slow, safe)
- check statement, drop proofs (fast cpu, fast io, totally unsafe wrt proofs)
- check statements, keep proof "thunks" (fast cpu, slow io, mostly safe wrt proofs when the thunks are forced, see `@check-proofs` below)

Currently Coq names the files `.vo`, `.vos` and `.vio` respectively but internally the format is the same (it is a segment based format, some segments are optional). Via `make` Coq supports various workflows, and indeed the different file names are motivated by the fact that this way different make rules are easily written/selected. The idea discussed at a Coq call was to support *most* of these workflows/backends by
- letting one chose a backend via `--profile` (or `(profile ..)`)
- implement dune specific Coq options to name the output .vo independently of the backend (so that dune rules can stay the same).

This is a WIP, and I don't have yet a Coq branch supporting the `-vos-dune` and `-vio-dune` flags I used there (but I'm pretty sure I can do this). I open the draft PR mainly to gather feedback and avoid duplication of work for #3486 . `make test-coq` seems to suggest I'm calling Coq with the expected flags in the blackbox test I've added.

Things to be checked by someone understanding dune internals:
- when the backend changes the .vo files need to be rebuilt (invalidated)
- eventually one can add a target `@check-proofs`. This target succeed immediately with the first backend, always fails with the second. With the third backend it runs a rule that depends on all .vo files and runs a specific coqc invocation (all proofs can be checked in parallel)~

Fix #3486 
CC @Zimmi48 @ejgallego

### CODE REVIEW

**Comment by ejgallego** [comment]
Thanks for the PR @gares, a couple of comments:

- it seems you don't intend to change the rules, then wouldn't it better to just implement `(coq (flags ...))` in the profile, and have users just put the flag there?
- if you don't change the build rules, you don't gain a lot, right?
- Wouldn't reusing `.vo` be confusing for the users too? IMHO it is better if we try to be consistent with the current way things are built.
- indeed if you change the set of flags, Dune will invalidate the targets.

Once we agree on the first point I will make a pass on the rest of the file.

**Comment by ejgallego** [comment]
Note that it is easy for example to introduce an alias for all `vio` / `vos` in a library, such as `@library/interfaces` [that would compile for example `.vos`].

An advantage of using different extensions is that I think incremental behavior is a bit better, for example if you decide to compile some part of you workspace with quick, this will force a full rebuild.

The conditional dependency on a `.vio` / `.vo` can be actually captured pretty well I think, tho you may want to need some flag at some point, or just piggyback on the release profile to disable rules that may produce unsound type-checking.

File: src/dune/coq_rules.ml
Comment (ejgallego) [inline src/dune/coq_rules.ml:103]: Small note, I will move this to `Coq_flags` soon as indeed handling of flags is becoming quite interesting.

**Comment by gares** [comment]
> * it seems you don't intend to change the rules, then wouldn't it better to just implement `(coq (flags ...))` in the profile, and have users just put the flag there?

that would work as well, but asking one to write `-vio-dune` seems less declarative to me.
for experimenting it is perfectly fine.

> * if you don't change the build rules, you don't gain a lot, right?

of course you gain: even if the dag of dependencies is the same, `coqc` is much faster in vos or vio mode since it does not check proofs.

> * Wouldn't reusing `.vo` be confusing for the users too? IMHO it is better if we try to be consistent with the current way things are built.

no, because in dune they don't see the build artifacts

> * indeed if you change the set of flags, Dune will invalidate the targets.

great!

File: src/dune/super_context.mli
Comment (ejgallego) [inline src/dune/super_context.mli:83]: Need to have a closer look but the part related to `env` / SC seems OK to me.

**Comment by ejgallego** [comment]
Also, DCO is needed, it is also recommended to run `make fmt` once the PR is ready, but we do this just before merging usually.

**Comment by gares** [comment]
Is the DCO just about adding a sign-of line to commits as CI reports?

**Comment by ejgallego** [comment]
> that would work as well, but asking one to write `-vio-dune` seems less declarative to me.
> for experimenting it is perfectly fine.

I'd say let's add a general `(coq (flags ...))` field for now, that may have other uses and it is fine indeed to keep supporting for a long time; a deeper change such a new `quick` profile may require more thinking, I'm not very convinced of adding a full new meaning for what amounts to setting a flag.

> of course you gain: even if the dag of dependencies is the same, coqc is much faster in vos or vio mode since it does not check proofs.

Indeed that's quite a bit, but I mean you can do this already using `(include ...)` in the flags field; ideally you would also get best of both worlds and `vio -> vo` etc...

> no, because in dune they don't see the build artifacts

Still people may want to look at _build to see what's going on, also this strategy of reusing names has a risk which is that people could install wrongly named libraries. Actually, this method won't produce installable packages that are both using `vos/vo` right?

Yup, the DCO is about signing-off your commits [which implies you agree with the contributing agreement]

More things needed: update documentation.

**Comment by gares** [comment]
> An advantage of using different extensions is that I think incremental behavior is a bit better, for example if you decide to compile some part of you workspace with quick, this will force a full rebuild.

This is one thing I would avoid as it has always been a mess in Coq.
Do you remember the "heuristic" about loading the .vio, or maybe the .vo if more recent... and now we also have vos/vok with the empty file hack. What I proposed at the Coq call is to rethink this, and start from making the 3 worlds disconnected, as profiles are if I'm not mistaken.

**Comment by ejgallego** [comment]
> Do you remember the "heuristic" about loading the .vio, or maybe the .vo if more recent... and now we also have vos/vok with the empty file hack. What I proposed at the Coq call is to question this choice, and start from the simpler approach of making the 3 worlds disconnected, as profiles are if I'm not mistaken.

Yup, but indeed these worlds are quite disconnected IIUC, for example vos are supposed to be interfaces, so I'm not sure the same file name would work as they are _too_ disconnected. Moreover, it seems this would prevent some valid use cases as I mention w.r.t. incremental setup.

I think it is OK to have interesting rules as long as we design them properly. Indeed that may mean we need to rethink the way `vos/vio` interact with profiles, but on the other hand this seems less radical than just bump all of them together.

**Comment by gares** [comment]
> Still people may want to look at _build to see what's going on, also this strategy of reusing names has a risk which is that people could install wrongly named libraries.

I see no difference with using dev or release on ocaml file. You look in there and you see a .cmx. Was it compiler with -opaque? You don't see it from the file name.

I also never feel the need to look into _build when I develop OCaml code (unless I'm debugging the build system, which is not something we expect our users to do often). Why would it be different for Coq?

> Actually, this method won't produce installable packages that are both using `vos/vo` right?

Having "both" serve no purpose, you only want the most complete one. In the approach I propose, it is always called vo. Print Assumption & Co can be updated to tell you that you are loading incomplete files.

Anyway, being able to to pass flags to coqc via profile is a good feature anyway, so maybe we should focus on that here.

**Comment by gares** [comment]
I've re targeted the PR to just support coq.flags. It works for me as well and requires no discussion about the design of alternative Coq backends.

**Review by ejgallego — CHANGES_REQUESTED** [review]
Looks pretty good, some tweaks are possible. @rgrinberg how does this look to you?

While I finish the review, this is missing:
- changelog entry
- documentation update.

**Comment by gares** [comment]
ping

File: src/dune/dune_env.ml
Comment (rgrinberg) [inline src/dune/dune_env.ml:177]: Should be 2.7 now.

File: src/dune/dune_env.mli
Comment (rgrinberg) [inline src/dune/dune_env.mli:29]: We don't have type aliases for other env flag fields, so let's keep it consistent and not introduce them here.

**Review by rgrinberg — APPROVED** [review]
Looks good and should be good to go for 2.7.

**Comment by ejgallego** [comment]
Thanks @rgrinberg ; @gares this looks good then, I will merge once the two comments are addressed.

**Comment by gares** [comment]
should be OK now

File: doc/dune-files.rst
Comment (ejgallego) [inline doc/dune-files.rst:1350]: Nit: we could indeed also mention in the `(coq.theory ...)` documentation that flags can be set in `env`

**Review by ejgallego — APPROVED** [review]

**Pull request merged**
